dids/src/main/kotlin/web5/sdk/dids/DidIonManager.kt

```diff
@@ -0,0 +1,353 @@
+
+private const val maxVerificationMethodIdLength = 50
+
+/**
+ * Base class for managing DID Ion operations. Uses the given [configuration].
+ */
+public sealed class DidIonManager(
+  private val configuration: DidIonConfiguration
+) : DidMethod<DidIonHandle, CreateDidIonOptions> {
+
+  private val mapper = jacksonObjectMapper()
+
+  private val operationsEndpoint = configuration.ionHost + operationsPath
+  private val identifiersEndpoint = configuration.ionHost + identifiersPath
+
+  private val engine: HttpClientEngine = if (configuration.engine == null) {
+    CIO.create {}
+  } else {
+    configuration.engine!!
+  }
+
+  private val client = HttpClient(engine) {
+    install(ContentNegotiation) {
+      jackson { mapper }
+    }
+  }
+
+  override val methodName: String = "ion"
+
+  /**
+   * Creates a [DidIonHandle], which includes a DID and it's associated DID Document. In order to ensure the creation
+   * works appropriately, the DID is resolved immediately after it's created.
+   *
+   * Note: [options] must be of type [CreateDidIonOptions].
+   * @throws [ResolutionException] When there is an error after resolution.
+   * @throws [InvalidStatusException] When any of the network requests return an invalid HTTP status code.
+   * @see [DidMethod.create] for details of each parameter.
+   */
+  override fun create(keyManager: KeyManager, options: CreateDidIonOptions?): DidIonHandle {
+    val (createOp, keys) = createOperation(keyManager, options)
+
+    val shortFormDidSegment =
+      Base64URL.encode(
+        Multihash.sum(Multicodec.SHA2_256, canonicalized(createOp.suffixData)).get()?.bytes()
+      ).toString()
+    val initialState = InitialState(
+      suffixData = createOp.suffixData,
+      delta = createOp.delta,
+    )
+    val longFormDidSegment = didUriSegment(initialState)
+
+    val response: HttpResponse = runBlocking {
+      client.post(operationsEndpoint) {
+        contentType(ContentType.Application.Json)
+        setBody(createOp)
+      }
+    }
+
+    val opBody = runBlocking {
+      response.bodyAsText()
+    }
+
+    if (response.status.isSuccess()) {
+      val shortFormDid = "did:ion:$shortFormDidSegment"
+      val longFormDid = "$shortFormDid:$longFormDidSegment"
+      val resolutionResult = resolve(longFormDid)
+
+      if (!resolutionResult.didResolutionMetadata?.error.isNullOrEmpty()) {
+        throw ResolutionException(
+          "error when resolving after creation: ${resolutionResult.didResolutionMetadata?.error}"
+        )
+      }
+
+      return DidIonHandle(
+        resolutionResult.didDocument.id.toString(),
+        keyManager,
+        IonCreationMetadata(
+          createOp,
+          shortFormDid,
+          longFormDid,
+          opBody,
+          keys
+        )
+      )
+    }
+    throw InvalidStatusException("received error response '$opBody'")
+  }
+
+  private inline fun <reified T> canonicalized(data: T): ByteArray {
```

Comment:
nearly impossible to trace down exceptions when using `inline` functions. Also `reified T` does not work well with Java at all. you're going to end up with just `Object`

Reply:
oh right good call! yep yep